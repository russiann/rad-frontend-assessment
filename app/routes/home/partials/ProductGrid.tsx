import React from "react";
import { Button, Card, Image, Badge, Chip } from "@heroui/react";
import { Icon } from "@iconify/react";
import { Link } from "react-router";
import { showPriceUpdateToast } from "../../../utils/notifications";
import type { Product } from "../../../utils/trpc";
import { cn } from "@heroui/react";
import { useChatContext } from "../../../contexts/ChatContext";

type ProductGridProps = {
  products: Product[];
  onAddToCart: (productName: string, productId: number) => void;
};

export function ProductGrid({ products, onAddToCart }: ProductGridProps) {
  const { isOpen: isChatOpen } = useChatContext();
  
  // Adaptive grid classes based on chat state and screen size
  const getGridClasses = () => {
    if (isChatOpen) {
      // When chat is open, use smaller minimum card width for better fitting
      return "grid gap-6";
    } else {
      // When chat is closed, allow more columns
      return "grid gap-6";
    }
  };

  const getGridStyle = () => {
    if (isChatOpen) {
      // When chat is open, minimum card width is larger to ensure good proportions
      return { 
        gridTemplateColumns: 'repeat(auto-fill, minmax(280px, 1fr))'
      };
    } else {
      // When chat is closed, allow smaller minimum width for more columns
      return { 
        gridTemplateColumns: 'repeat(auto-fill, minmax(250px, 1fr))'
      };
    }
  };

  return (
    <div className={getGridClasses()} style={getGridStyle()}>
      {products.map((product) => {
        const isAvailable = product.isAvailable ?? true;
        
        return (
          <Card key={product.id} className="relative">
            {product.id === 3 && (
              <Badge 
                content="SALE" 
                color="danger" 
                placement="top-right"
                className="z-10 m-2 cursor-pointer"
                onClick={() => showPriceUpdateToast(product.name, product.id)}
              >
                <span></span>
              </Badge>
            )}
            
            {/* Availability indicator */}
            {!product.isAvailable && (
              <Chip 
                size="sm" 
                color="danger" 
                variant="flat"
                className="absolute top-2 left-2 z-10"
              >
                Unavailable
              </Chip>
            )}
            
            <Link to={`/product/${product.id}`}>
              <div className={`relative ${!isAvailable ? 'opacity-75' : ''}`}>
                <Image
                  removeWrapper
                  alt={product.name}
                  className="w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
                  src={product.image}
                />
                {!isAvailable && (
                  <div className="absolute inset-0 bg-black bg-opacity-30 flex items-center justify-center">
                    <Icon icon="lucide:x-circle" className="text-white text-4xl" />
                  </div>
                )}
              </div>
            </Link>
            
            <div className="p-4 space-y-3">
              <Link to={`/product/${product.id}`} className="block">
                <h3 className={`font-semibold text-lg hover:text-primary transition-colors ${!isAvailable ? 'text-default-400' : ''}`}>
                  {product.name}
                </h3>
              </Link>
              <div className="flex items-center justify-between">
                <p className={`font-medium ${!isAvailable ? 'text-default-400' : 'text-default-600'}`}>
                  ${product.price.toFixed(2)}
                </p>
                <p className="text-xs text-default-500">{product.category}</p>
              </div>
              <Button 
                className={cn(
                  "w-full",
                  !isAvailable && "opacity-50 cursor-not-allowed"
                )}
                color="primary"
                variant="flat"
                size="sm"
                onPress={() => isAvailable && onAddToCart(product.name, product.id)}
                isDisabled={!isAvailable}
              >
                {isAvailable ? "Add to Cart" : "Unavailable"}
              </Button>
            </div>
          </Card>
        );
      })}
    </div>
  );
} 