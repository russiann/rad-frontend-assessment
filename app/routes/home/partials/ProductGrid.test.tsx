import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ProductGrid } from './ProductGrid';
import type { Product } from '../../../utils/trpc';

// Mock dependencies
vi.mock('../../../utils/notifications', () => ({
  showPriceUpdateToast: vi.fn(),
}));

vi.mock('../../../contexts/ChatContext', () => ({
  useChatContext: () => ({
    isOpen: false,
  }),
}));

// Mock react-router Link
vi.mock('react-router', () => ({
  Link: ({ children, to }: { children: React.ReactNode; to: string }) => (
    <a href={to}>{children}</a>
  ),
}));

describe('ProductGrid', () => {
  const mockOnAddToCart = vi.fn();

  const mockProducts: Product[] = [
    {
      id: 1,
      name: 'Gaming Laptop',
      price: 999.99,
      category: 'Electronics',
      description: 'High-performance gaming laptop',
      image: '/laptop.jpg',
      isAvailable: true,
      stock: 10,
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
    {
      id: 2,
      name: 'Cotton T-Shirt',
      price: 29.99,
      category: 'Clothing',
      description: 'Comfortable cotton t-shirt',
      image: '/tshirt.jpg',
      isAvailable: true,
      stock: 5,
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
    {
      id: 3,
      name: 'Running Shoes',
      price: 79.99,
      category: 'Footwear',
      description: 'Lightweight running shoes',
      image: '/shoes.jpg',
      isAvailable: false,
      stock: 0,
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  ];

  const defaultProps = {
    products: mockProducts,
    onAddToCart: mockOnAddToCart,
  };

  beforeEach(() => {
    mockOnAddToCart.mockClear();
    vi.clearAllMocks();
  });

  describe('Basic Rendering', () => {
    it('should render all products', () => {
      render(<ProductGrid {...defaultProps} />);

      expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
      expect(screen.getByText('Cotton T-Shirt')).toBeInTheDocument();
      expect(screen.getByText('Running Shoes')).toBeInTheDocument();
    });

    it('should display product prices correctly', () => {
      render(<ProductGrid {...defaultProps} />);

      expect(screen.getByText('$999.99')).toBeInTheDocument();
      expect(screen.getByText('$29.99')).toBeInTheDocument();
      expect(screen.getByText('$79.99')).toBeInTheDocument();
    });

    it('should display product categories', () => {
      render(<ProductGrid {...defaultProps} />);

      expect(screen.getByText('Electronics')).toBeInTheDocument();
      expect(screen.getByText('Clothing')).toBeInTheDocument();
      expect(screen.getByText('Footwear')).toBeInTheDocument();
    });

    it('should render product links correctly', () => {
      render(<ProductGrid {...defaultProps} />);

      const laptopLinks = screen.getAllByRole('link');
      const productLink = laptopLinks.find(link => 
        link.getAttribute('href') === '/product/1'
      );
      expect(productLink).toBeInTheDocument();
    });
  });

  describe('Product Availability', () => {
    it('should show "Add to Cart" for available products', () => {
      render(<ProductGrid {...defaultProps} />);

      const addToCartButtons = screen.getAllByText('Add to Cart');
      expect(addToCartButtons).toHaveLength(2); // Only available products
    });

    it('should show "Unavailable" for unavailable products', () => {
      render(<ProductGrid {...defaultProps} />);

      expect(screen.getAllByText('Unavailable')).toHaveLength(2); // Chip + Button
      
      const unavailableButton = screen.getByRole('button', { name: /unavailable/i });
      expect(unavailableButton).toBeDisabled();
    });

    it('should show unavailable chip for unavailable products', () => {
      render(<ProductGrid {...defaultProps} />);

      // Check for the unavailable chip/badge
      const unavailableElements = screen.getAllByText('Unavailable');
      expect(unavailableElements).toHaveLength(2); // Chip + Button
      expect(unavailableElements[0]).toBeInTheDocument();
    });
  });

  describe('Add to Cart Functionality', () => {
    it('should call onAddToCart when clicking available product button', async () => {
      const user = userEvent.setup();
      render(<ProductGrid {...defaultProps} />);

      const addToCartButtons = screen.getAllByText('Add to Cart');
      await user.click(addToCartButtons[0]);

      expect(mockOnAddToCart).toHaveBeenCalledWith('Gaming Laptop', 1);
    });

    it('should not call onAddToCart for unavailable products', async () => {
      const user = userEvent.setup();
      render(<ProductGrid {...defaultProps} />);

      const unavailableButton = screen.getByRole('button', { name: /unavailable/i });
      
      // Try to click but it shouldn't work since it's disabled
      await user.click(unavailableButton);
      expect(mockOnAddToCart).not.toHaveBeenCalled();
    });
  });

  describe('Special Product Features', () => {
    it('should show SALE badge for product ID 3', () => {
      const productsWithSale = [
        {
          id: 3,
          name: 'Sale Item',
          price: 99.99,
          category: 'Electronics',
          description: 'On sale item',
          image: '/sale.jpg',
          isAvailable: true,
          stock: 10,
          createdAt: '2024-01-01T00:00:00Z',
          updatedAt: '2024-01-01T00:00:00Z',
        },
      ];

      render(<ProductGrid products={productsWithSale} onAddToCart={mockOnAddToCart} />);

      expect(screen.getByText('SALE')).toBeInTheDocument();
    });

    it('should call price update toast when clicking SALE badge', async () => {
      const user = userEvent.setup();
      const productsWithSale = [
        {
          id: 3,
          name: 'Sale Item',
          price: 99.99,
          category: 'Electronics',
          description: 'On sale item',
          image: '/sale.jpg',
          isAvailable: true,
          stock: 10,
          createdAt: '2024-01-01T00:00:00Z',
          updatedAt: '2024-01-01T00:00:00Z',
        },
      ];

      const { showPriceUpdateToast } = await import('../../../utils/notifications');

      render(<ProductGrid products={productsWithSale} onAddToCart={mockOnAddToCart} />);

      const saleBadge = screen.getByText('SALE');
      await user.click(saleBadge);

      expect(showPriceUpdateToast).toHaveBeenCalledWith('Sale Item', 3);
    });
  });

  describe('Empty State', () => {
    it('should render empty grid when no products', () => {
      render(<ProductGrid products={[]} onAddToCart={mockOnAddToCart} />);

      expect(screen.queryByText('Gaming Laptop')).not.toBeInTheDocument();
      expect(screen.queryByText('Add to Cart')).not.toBeInTheDocument();
    });
  });

  describe('Chat Context Integration', () => {
    it('should adjust grid layout when chat is open', () => {
      // Mock chat as open
      vi.doMock('../../../contexts/ChatContext', () => ({
        useChatContext: () => ({
          isOpen: true,
        }),
      }));

      render(<ProductGrid {...defaultProps} />);

      // The grid should still render products regardless of chat state
      expect(screen.getByText('Gaming Laptop')).toBeInTheDocument();
    });
  });

  describe('snapshots', () => {
    it('renders product grid correctly', () => {
      const { container } = render(<ProductGrid {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders with empty products array', () => {
      const { container } = render(
        <ProductGrid products={[]} onAddToCart={mockOnAddToCart} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders with mix of available and unavailable products', () => {
      const { container } = render(<ProductGrid {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});