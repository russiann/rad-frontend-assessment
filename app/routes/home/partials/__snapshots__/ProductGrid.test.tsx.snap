// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductGrid > snapshots > renders product grid correctly 1`] = `
<div
  class="grid gap-6"
  style="grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));"
>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <a
      href="/product/1"
    >
      <div
        class="relative "
      >
        <img
          alt="Gaming Laptop"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/laptop.jpg"
        />
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/1"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors "
        >
          Gaming Laptop
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-600"
        >
          $
          999.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Electronics
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <a
      href="/product/2"
    >
      <div
        class="relative "
      >
        <img
          alt="Cotton T-Shirt"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/tshirt.jpg"
        />
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/2"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors "
        >
          Cotton T-Shirt
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-600"
        >
          $
          29.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Clothing
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <div
      class="relative inline-flex shrink-0"
    >
      <span />
      <span
        class="flex flex-wrap absolute box-border rounded-full whitespace-nowrap place-content-center origin-center items-center select-none font-regular scale-100 opacity-100 subpixel-antialiased data-[invisible=true]:scale-0 data-[invisible=true]:opacity-0 px-1 text-small transition-transform-opacity !ease-soft-spring !duration-300 border-2 border-background bg-danger text-danger-foreground top-[5%] right-[5%] translate-x-1/2 -translate-y-1/2 z-10 m-2 cursor-pointer"
      >
        SALE
      </span>
    </div>
    <div
      class="max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-6 text-tiny rounded-full bg-danger/20 text-danger-600 dark:text-danger-500 absolute top-2 left-2 z-10"
    >
      <span
        class="flex-1 text-inherit font-normal px-1"
      >
        Unavailable
      </span>
    </div>
    <a
      href="/product/3"
    >
      <div
        class="relative opacity-75"
      >
        <img
          alt="Running Shoes"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/shoes.jpg"
        />
        <div
          class="absolute inset-0 bg-black bg-opacity-30 flex items-center justify-center"
        >
          <span />
        </div>
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/3"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors text-default-400"
        >
          Running Shoes
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-400"
        >
          $
          79.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Footwear
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small pointer-events-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full opacity-50 cursor-not-allowed"
        data-disabled="true"
        data-react-aria-pressable="true"
        disabled=""
        type="button"
      >
        Unavailable
      </button>
    </div>
  </div>
</div>
`;

exports[`ProductGrid > snapshots > renders with empty products array 1`] = `
<div
  class="grid gap-6"
  style="grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));"
/>
`;

exports[`ProductGrid > snapshots > renders with mix of available and unavailable products 1`] = `
<div
  class="grid gap-6"
  style="grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));"
>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <a
      href="/product/1"
    >
      <div
        class="relative "
      >
        <img
          alt="Gaming Laptop"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/laptop.jpg"
        />
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/1"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors "
        >
          Gaming Laptop
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-600"
        >
          $
          999.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Electronics
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <a
      href="/product/2"
    >
      <div
        class="relative "
      >
        <img
          alt="Cotton T-Shirt"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/tshirt.jpg"
        />
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/2"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors "
        >
          Cotton T-Shirt
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-600"
        >
          $
          29.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Clothing
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        Add to Cart
      </button>
    </div>
  </div>
  <div
    class="flex flex-col overflow-hidden h-auto text-foreground box-border bg-content1 outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none relative"
    tabindex="-1"
  >
    <div
      class="relative inline-flex shrink-0"
    >
      <span />
      <span
        class="flex flex-wrap absolute box-border rounded-full whitespace-nowrap place-content-center origin-center items-center select-none font-regular scale-100 opacity-100 subpixel-antialiased data-[invisible=true]:scale-0 data-[invisible=true]:opacity-0 px-1 text-small transition-transform-opacity !ease-soft-spring !duration-300 border-2 border-background bg-danger text-danger-foreground top-[5%] right-[5%] translate-x-1/2 -translate-y-1/2 z-10 m-2 cursor-pointer"
      >
        SALE
      </span>
    </div>
    <div
      class="max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-6 text-tiny rounded-full bg-danger/20 text-danger-600 dark:text-danger-500 absolute top-2 left-2 z-10"
    >
      <span
        class="flex-1 text-inherit font-normal px-1"
      >
        Unavailable
      </span>
    </div>
    <a
      href="/product/3"
    >
      <div
        class="relative opacity-75"
      >
        <img
          alt="Running Shoes"
          class="relative z-10 shadow-black/5 data-[loaded=true]:opacity-100 shadow-none opacity-0 motion-reduce:transition-none !duration-300 rounded-large w-full h-60 object-cover cursor-pointer transition-transform hover:scale-[1.03]"
          src="/shoes.jpg"
        />
        <div
          class="absolute inset-0 bg-black bg-opacity-30 flex items-center justify-center"
        >
          <span />
        </div>
      </div>
    </a>
    <div
      class="p-4 space-y-3"
    >
      <a
        href="/product/3"
      >
        <h3
          class="font-semibold text-lg hover:text-primary transition-colors text-default-400"
        >
          Running Shoes
        </h3>
      </a>
      <div
        class="flex items-center justify-between"
      >
        <p
          class="font-medium text-default-400"
        >
          $
          79.99
        </p>
        <p
          class="text-xs text-default-500"
        >
          Footwear
        </p>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-3 min-w-16 h-8 text-tiny gap-2 rounded-small pointer-events-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-primary/20 text-primary-600 data-[hover=true]:opacity-hover w-full opacity-50 cursor-not-allowed"
        data-disabled="true"
        data-react-aria-pressable="true"
        disabled=""
        type="button"
      >
        Unavailable
      </button>
    </div>
  </div>
</div>
`;
